components/range_cache_memory_engine/src/range_stats.rs

```diff
@@ -381,6 +337,98 @@ impl RangeStatsManager {
         };
         (regions_to_load, regions_to_evict)
     }
+
+    // Evict regions with less read flow until the memory usage is below the soft
+    // limit threshold.
+    pub(crate) fn evict_on_soft_limit_reached<F>(
+        &self,
+        mut evict_region: F,
+        delete_range_scheduler: &Scheduler<BackgroundTask>,
+        cached_region_ids: Vec<u64>,
+        memory_controller: &MemoryController,
+    ) where
+        F: FnMut(
+            &CacheRegion,
+            EvictReason,
+            Option<Box<dyn FnOnce() + Send + Sync>>,
+        ) -> Vec<CacheRegion>,
+    {
+        // Get regions' stat of the cached region and sort them by next + prev in
+        // descending order.
+        let regions_activity = self
+            .info_provider
+            .get_regions_stat(cached_region_ids.clone())
+            .unwrap();
+        if regions_activity.is_empty() {
+            return;
+        }
+
+        // Use the average mvcc amplification to filter the regions with high mvcc
+        // amplification
+        let avg_mvcc_amplification = regions_activity
+            .iter()
+            .map(|(_, ra)| ra.cop_detail.mvcc_amplification())
+            .sum::<f64>()
+            / regions_activity.len() as f64;
+        let evict_candidates: Vec<_> = {
+            let mut regions_loaded = self.region_loaded_at.write().unwrap();
+            let mut evict_candidates = regions_activity;
+            evict_candidates.retain(|(r, ra)| {
+                let time_loaded = regions_loaded.entry(r.id).or_insert(Instant::now());
+                ra.cop_detail.mvcc_amplification() < avg_mvcc_amplification
+                    && Instant::now() - *time_loaded > self.evict_min_duration
+            });
+            evict_candidates.sort_by(|a, b| {
+                let next_prev_a = a.1.cop_detail.iterated_count();
+                let next_prev_b = b.1.cop_detail.iterated_count();
+                next_prev_a.cmp(&next_prev_b)
+            });
+
+            evict_candidates.into_iter().map(|(r, _)| r).collect()
+        };
+        // Evict two regions each time to reduce the probability that an un-dropped
+        // ongoing snapshot blocks the process
+        for regions in evict_candidates.chunks(2) {
+            let mut rxs = vec![];
```

Comment:
No need to build multiple channel, you can share the same channel with capacity 2 and receive twice.

Reply:
Good suggestion.